library/net.cpp

```diff
@@ -63,27 +65,162 @@ enum dnet_socket_state {
+		dnet_net_state *st;
+		int s;
+		int err;
+
+		st = dnet_state_search_by_addr(node, address);
+		if (st) {
+			err = -EEXIST;
+
+			dnet_log(node, DNET_LOG_NOTICE, "Address %s already exists in route table",
+				 dnet_addr_string(address));
+			dnet_state_put(st);
+			return err;
+		}
+
+		salen = address->addr_len;
+		sa = (sockaddr *)address;
+
+		sa->sa_family = address->family;
+
+		s = socket(sa->sa_family, SOCK_STREAM, IPPROTO_TCP);
+		if (s < 0) {
+			err = -errno;
+
+			dnet_log_err(node, "Failed to create socket for %s: family: %d",
+				     dnet_addr_string(address), sa->sa_family);
+			return err;
+		}
+
+		fcntl(s, F_SETFL, O_NONBLOCK);
+		fcntl(s, F_SETFD, FD_CLOEXEC);
+
+		if (listening) {
+			err = 1;
+			setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &err, 4);
+
+			err = bind(s, sa, salen);
+			if (err) {
+				err = -errno;
+				dnet_log_err(node, "Failed to bind to %s",
+					     dnet_addr_string(address));
+				return err;
+			}
+
+			err = listen(s, 10240);
+			if (err) {
+				err = -errno;
+				dnet_log_err(node, "Failed to listen at %s",
+					     dnet_addr_string(address));
+				return err;
+			}
+
+			dnet_log(node, DNET_LOG_INFO, "Server is now listening at %s.",
+				 dnet_addr_string(address));
+		} else {
+			dnet_log(node, DNET_LOG_INFO, "Added %s to connect list",
+				 dnet_addr_string(address));
+		}
+
+		return s;
+	}
+
+	void close() {
+		if (s >= 0) {
+			::close(s);
+			s = -1;
+		}
+	}
+
 	int s;
 	int ok;
+	dnet_addr addr;
 	dnet_socket_state state;
 	dnet_cmd io_cmd;
-	void *buffer;
-	void *io_data;
+	std::unique_ptr<char[]> buffer;
+	char *io_data;
 	size_t io_size;
 	int version[4];
 	bool ask_route_list;
 };
 
-struct dnet_addr_socket_list
+typedef std::shared_ptr<dnet_addr_socket> dnet_addr_socket_ptr;
+
+static inline bool operator <(const dnet_addr &first, const dnet_addr &second)
 {
-	list_head entry;
-	size_t sockets_count;
-	dnet_addr_socket sockets[0];
+	return dnet_addr_cmp(&first, &second) < 0;
+}
+
+static inline bool operator ==(const dnet_addr &first, const dnet_addr &second)
+{
+	return dnet_addr_equal(&first, &second);
+}
+
+struct dnet_addr_socket_comparator
+{
+	bool operator () (const dnet_addr_socket_ptr &first, const dnet_addr_socket_ptr &second) {
+		return first->addr < second->addr;
+	}
 };
 
+typedef std::set<dnet_addr_socket_ptr, dnet_addr_socket_comparator> dnet_addr_socket_set;
+
 struct dnet_connect_state
 {
+	dnet_connect_state()
+	: lock_inited(false),
+	 node(nullptr),
+	 epollfd(-1),
+	 interruptfd(-1),
+	 failed_count(0),
+	 succeed_count(0),
+	 total_count(0),
+	 finished(false)
+	{
+		atomic_set(&refcnt, 0);
+		atomic_set(&route_list_count, 0);
+	}
+
+	~dnet_connect_state() {
+		if (lock_inited)
+			pthread_mutex_destroy(&lock);
+		if (epollfd >= 0)
+			close(epollfd);
+		if (interruptfd >= 0)
+			close(interruptfd);
+	}
+
+	dnet_connect_state *get()
+	{
+		atomic_add(&refcnt, 1);
+		return this;
+	}
+
+	void put()
+	{
+		if (atomic_dec_and_test(&refcnt))
+			delete this;
```

Comment:
What are the benefits of using dnet_connect_state::put function instead of dnet_connect_state_put? As I see first one only brings confusion and doesn't provide any benefits at all